ContextFinder falls back to context classloader resulting in class cast errors

When we call `JAXBContext.newInstance(Class<?>...)` or `JAXBContext.newInstance(Class<?>[], Map<String,?>)` to create a new `JAXBContext` we can't pass in a classloader. This then in return means that `ContextFinder.newInstance` gets called with the context classloader.

Given that the context classloader can be "higher" in the chain of classloaders than the current class's classloader and that, since the jaxb-api is no longer part of java core, this might mean that the `JAXBContext` class has already been loaded (for instance, in the current class's classloader) in a way that makes it inaccessible to the current thread's context classloader.

This means that if we end up at line 362 and 363 of `ContextFinder.find(Class<?>[], Map<String, ?>)` and then eventually at[ line 230 ](https://github.com/eclipse-ee4j/jaxb-api/blob/797f8c8564df1dee944a0ca3ea3622d518e1ba71/jaxb-api/src/main/java/javax/xml/bind/ContextFinder.java#L230) where we will load the factory class using the context classloader 

`spi = ServiceLoaderUtil.safeLoadClass(className, ModuleUtil.DEFAULT_FACTORY_CLASS, getContextClassLoader());`

we might end up with a `JAXBContextImpl` whose class is loaded on the context clasloader and which implements `JAXBContext` whose class is loaded on the context classloader and which therefore is not an instance of the `JAXBContext` whose class got loaded on the current class's classloader!
This in return means that the instanceof check [on line 252 of ContextFinder](https://github.com/eclipse-ee4j/jaxb-api/blob/797f8c8564df1dee944a0ca3ea3622d518e1ba71/jaxb-api/src/main/java/javax/xml/bind/ContextFinder.java#L252) would fail since the returned JAXBContextImpl is not an instance of the `JAXBContext` the `ContextFinder` knows. 

I ran into this issue while using javax.xml.bind:jaxb-api:2.3.1 on JDK 11 (specifically, while trying to read a persistence.xml using the hibernate jpa metamodel generator through the maven processor plugin). Some more detail on the hibernate jira [HHH-13794](https://hibernate.atlassian.net/browse/HHH-13794) and in [my blog post](https://www.crashdata.nl/uncategorized/hibernate-metamodelgen-classcastexception-on-jaxbcontext).

I ended up with a stacktrace like:

> Caused by: javax.xml.bind.JAXBException: ClassCastException: attempting to cast jar:file:/home/haster/.m2/repository/javax/xml/bind/jaxb-api/2.3.1/jaxb-api-2.3.1.jar!/javax/xml/bind/JAXBContext.class to jar:file:/home/haster/.m2/repository/javax/xml/bind/jaxb-api/2.3.1/jaxb-api-2.3.1.jar!/javax/xml/bind/JAXBContext.class.  Please make sure that you are specifying the proper ClassLoader.    
>     at javax.xml.bind.ContextFinder.handleClassCastException(ContextFinder.java:157)
>     at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:300)
>     at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:286)
>     at javax.xml.bind.ContextFinder.find(ContextFinder.java:409)
>     at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:721)
>     at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:662)
>     at org.hibernate.jpamodelgen.util.xml.XmlParserHelper.getJaxbRoot(XmlParserHelper.java:122)
>     ... 39 more

I see no reason why this wouldn't fail with the current version of jaxb-api but I haven't yet been able to test it. I think this will fail any time where the JAXBContext as used in ContextFinder gets loaded on a classloader that is not the current thread's context classloader. 

Constructing a testcase is somewhat less trivial since you have to make sure that the classloaders are configured correctly. I'm fairly certain that the testcase I provided for HHH-13794](https://hibernate.atlassian.net/browse/HHH-13794) would work for (or be easily adapted to) the current jakarta jaxb-api.

Update: I tested the testcase in HHH-13794 against jakarta.xml.bind:jakarta.xml.bind-api:2.3.2 and the error was there as well.

This issue is similar to, though definitely not the same as, #99.

I propose two solutions to mitigate this issue:

1. Add an API method to construct a `JAXBContext` with a passed set of classes to be bound while also passing in a `ClassLoader` to be used to load the context. Something like `JAXBContext.newInstance(Class<?>[], Map<String,?>, ClassLoader)`.

2. Use a different "default classloader" instead of the current thread's context classloader for all such situations . Maybe just `getClass().getClassloader()` or `JAXBContext.class.getClassLoader()` . That last one would at the very least ensure that the instanceof check against `JAXBContext` succeeds.

Any solution here? 

I'm now starting migration from "javax.*" to the "jakarta.*" namespace and I'm getting:
```
Caused by: jakarta.xml.bind.JAXBException: ClassCastException: attempting to cast jar:file:/home/developer/.m2/repository/jakarta/xml/bind/jakarta.xml.bind-api/3.0.0/jakarta.xml.bind-api-3.0.0.jar!/jakarta/xml/bind/JAXBContext.class to jar:file:/home/developer/.m2/repository/jakarta/xml/bind/jakarta.xml.bind-api/3.0.0/jakarta.xml.bind-api-3.0.0.jar!/jakarta/xml/bind/JAXBContext.class.  Please make sure that you are specifying the proper ClassLoader.    
	at jakarta.xml.bind.ContextFinder.handleClassCastException(ContextFinder.java:111)
	at jakarta.xml.bind.ContextFinder.newInstance(ContextFinder.java:254)
	at jakarta.xml.bind.ContextFinder.newInstance(ContextFinder.java:240)
	at jakarta.xml.bind.ContextFinder.find(ContextFinder.java:375)
	at jakarta.xml.bind.JAXBContext.newInstance(JAXBContext.java:691)
	at jakarta.xml.bind.JAXBContext.newInstance(JAXBContext.java:632)
```

--- a/jaxb-api/src/main/java/jakarta/xml/bind/ContextFinder.java
+++ b/jaxb-api/src/main/java/jakarta/xml/bind/ContextFinder.java
@@ -239,11 +239,18 @@ class ContextFinder {
     /**
      * Create an instance of a class using the thread context ClassLoader
      */
-    static JAXBContext newInstance(Class[] classes, Map properties, String className) throws JAXBException {
+    private static JAXBContext newInstance(Class[] classes, Map properties, String className) throws JAXBException {
+        return newInstance(classes, properties, className, getContextClassLoader());
+    }
+
+    /**
+     * Create an instance of a class using passed in ClassLoader
+     */
+    private static JAXBContext newInstance(Class[] classes, Map properties, String className, ClassLoader loader) throws JAXBException {
 
         Class spi;
         try {
-            spi = ServiceLoaderUtil.safeLoadClass(className, DEFAULT_FACTORY_CLASS, getContextClassLoader());
+            spi = ServiceLoaderUtil.safeLoadClass(className, DEFAULT_FACTORY_CLASS, loader);
         } catch (ClassNotFoundException e) {
             throw new JAXBException(Messages.format(Messages.DEFAULT_PROVIDER_NOT_FOUND), e);
         }
@@ -358,7 +365,7 @@ class ContextFinder {
                                 jaxbPropertiesUrl,
                                 JAXBContext.JAXB_CONTEXT_FACTORY, JAXB_CONTEXT_FACTORY_DEPRECATED);
 
-                return newInstance(classes, properties, factoryClassName);
+                return newInstance(classes, properties, factoryClassName, c.getClassLoader());
             }
 
         }
@@ -375,8 +382,8 @@ class ContextFinder {
         }
 
         // to ensure backwards compatibility
-        String className = firstByServiceLoaderDeprecated(JAXBContext.class, getContextClassLoader());
-        if (className != null) return newInstance(classes, properties, className);
+        String className = firstByServiceLoaderDeprecated(JAXBContext.class, classes[0].getClassLoader());
+        if (className != null) return newInstance(classes, properties, className, classes[0].getClassLoader());
 
         logger.fine("Trying to create the platform default provider");
         Class ctxFactoryClass =

